src/mlpack/core/hpt/cv_function.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file cv_function.hpp
+ * @author Kirill Mishchenko
+ *
+ * A cross-validation wrapper for optimizers.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_HPT_CV_FUNCTION_HPP
+#define MLPACK_CORE_HPT_CV_FUNCTION_HPP
+
+#include <mlpack/core.hpp>
+
+namespace mlpack {
+namespace hpt {
+
+/**
+ * This wrapper serves for adapting the interface of the cross-validation
+ * classes to the one that can be utilized by the mlpack optimizers.
+ *
+ * This class is not supposed to be used directly by users. To tune
+ * hyper-parameters see HyperParameterTuner.
+ *
+ * @tparam CVType A cross-validation strategy.
+ * @tparam TotalArgs The total number of arguments that are supposed to be
+ *     passed to the Evaluate method of a CVType object.
+ * @tparam BoundArgs Types of arguments (wrapped into the BoundArg struct) that
+ *     should be passed into the Evaluate method of a CVType object but are not
+ *     going to be passed into the Evaluate method of a CVFunction object.
+ */
+template<typename CVType, size_t TotalArgs, typename... BoundArgs>
+class CVFunction
+{
+ public:
+  /**
+   * Initialize a CVFunction object.
+   *
+   * @param cv A cross-validation object.
+   * @param BoundArgs Arguments that should be passed into the Evaluate method
+   *     of the CVType object but are not going to be passed into the Evaluate
+   *     method of this object.
+   */
+  CVFunction(CVType& cv, const BoundArgs&... args);
+
+  /**
+   * Run cross-validation with the bound and passed parameters.
+   *
+   * @param parameters Arguments (rather than the bound arguments) that should
+   *     be passed into the Evaluate method of the CVType object.
+   */
+  double Evaluate(const arma::mat& parameters);
+
+  //! The used machine learning algorithm.
+  using MLAlgorithm = typename
+      std::remove_reference<decltype(std::declval<CVType>().Model())>::type;
```

Comment:
We can also make `HyperParameterTuner` pass the `MLAlgorithm` type to `CVFunction`. It will be simpler too, and without making additional requirements for `CVType`. What do you think?